planner/core/find_best_task.go

```diff
@@ -407,17 +420,163 @@ func (ds *DataSource) findBestTask(prop *property.PhysicalProperty) (t task, err
+			}
+			if idxTask.cost() < t.cost() {
+				t = idxTask
+			}
+			continue
+		}
+		idxMergeTask, err := ds.convertToIndexMergeScan(prop, candidate)
 		if err != nil {
 			return nil, err
 		}
-		if idxTask.cost() < t.cost() {
-			t = idxTask
+		if idxMergeTask.cost() < t.cost() {
+			t = idxMergeTask
 		}
 	}
 	return
 }
 
+func (ds *DataSource) convertToIndexMergeScan(prop *property.PhysicalProperty, candidate *candidatePath) (task task, err error) {
+	if prop.TaskTp == property.CopSingleReadTaskType {
+		return invalidTask, nil
+	}
+	if !prop.IsEmpty() && !candidate.isMatchProp {
+		return invalidTask, nil
+	}
+	path := candidate.path
+	totalRowCount := 0.0
+	partPaths := make([]PhysicalPlan, 0, len(path.partialIndexPaths))
+	cop := &copTask{}
+	for _, partPath := range path.partialIndexPaths {
+		if partPath.isTablePath {
+			ts := PhysicalTableScan{
+				Table:           ds.tableInfo,
+				Columns:         ds.Columns,
+				TableAsName:     ds.TableAsName,
+				DBName:          ds.DBName,
+				isPartition:     ds.isPartition,
+				physicalTableID: ds.physicalTableID,
+			}.Init(ds.ctx)
+			ts.SetSchema(ds.schema)
+			if ts.Table.PKIsHandle {
+				if pkColInfo := ts.Table.GetPkColInfo(); pkColInfo != nil {
+					if ds.statisticTable.Columns[pkColInfo.ID] != nil {
+						ts.Hist = &ds.statisticTable.Columns[pkColInfo.ID].Histogram
+					}
+				}
+			}
+			path := candidate.path
+			ts.Ranges = path.ranges
+			ts.AccessCondition, ts.filterCondition = path.accessConds, path.tableFilters
+			rowCount := path.countAfterAccess
+			ts.stats = property.NewSimpleStats(rowCount)
+			ts.stats.StatsVersion = ds.statisticTable.Version
+			if ds.statisticTable.Pseudo {
+				ts.stats.StatsVersion = statistics.PseudoVersion
+			}
+			totalRowCount += rowCount
+			partPaths = append(partPaths, ts)
+			continue
+		}
+		idx := partPath.index
+		is := PhysicalIndexScan{
+			Table:            ds.tableInfo,
+			TableAsName:      ds.TableAsName,
+			DBName:           ds.DBName,
+			Columns:          ds.Columns,
+			Index:            idx,
+			IdxCols:          path.idxCols,
+			IdxColLens:       path.idxColLens,
+			AccessCondition:  path.accessConds,
+			Ranges:           path.ranges,
+			dataSourceSchema: ds.schema,
+			isPartition:      ds.isPartition,
+			physicalTableID:  ds.physicalTableID,
+		}.Init(ds.ctx)
+		statsTbl := ds.statisticTable
+		if statsTbl.Indices[idx.ID] != nil {
+			is.Hist = &statsTbl.Indices[idx.ID].Histogram
+		}
+		is.initSchema(ds.id, idx, true)
+		rowCount := partPath.countAfterAccess
+		if prop.IsEmpty() && prop.ExpectedCnt < ds.stats.RowCount {
+			selectivity := ds.stats.RowCount / path.countAfterAccess
+			rowCount = math.Min(prop.ExpectedCnt/selectivity, rowCount)
+		}
+		is.stats = property.NewSimpleStats(rowCount)
+		is.stats.StatsVersion = ds.statisticTable.Version
+		if ds.statisticTable.Pseudo {
+			is.stats.StatsVersion = statistics.PseudoVersion
+		}
+		totalRowCount += rowCount
+		partPaths = append(partPaths, is)
```

Comment:
We may have `indexFilters` on this partPath, so we may need additional `PhysicalSelection` on top of this index scan.